### ISSUE TITLE
5.0.3 upgrade sql v6.0

### ISSUE THREAD

**Issue by sjpadgett** [openemr/openemr#3885]
### Describe the bug
First pass hung up towards end.
Second pass gave below.
![image](https://user-images.githubusercontent.com/1859985/91188123-f06be900-e6be-11ea-977c-660881804528.png)

Third pass finished.

This is on my development database i've carried around since v4.1 so could be a fluke or bares watching.

**Comment by sjpadgett** [comment]
502 to v6.0
hangs here:
![image](https://user-images.githubusercontent.com/1859985/91367078-842ed980-e7d3-11ea-8ead-eb82971e4d91.png)
then
[26-Aug-2020 19:54:23 America/New_York] SQL Error with statement:query failed: CREATE UNIQUE INDEX `uuid` ON `form_encounter` (`uuid`)--Duplicate key name &#039;uuid&#039;==&gt;C:\xampp\htdocs\openemr-6\library\sql_upgrade_fx.php at 1074:sqlStatement
[26-Aug-2020 19:54:28 America/New_York] SQL Error with statement:query failed: ALTER TABLE `form_encounter` ADD `class_code` VARCHAR(10) NOT NULL  DEFAULT &quot;AMB&quot;--Duplicate column name &#039;class_code&#039;==&gt;C:\xampp\htdocs\openemr-6\library\sql_upgrade_fx.php at 1074:sqlStatement

I can't tell what is breaking. Appears to make mods in DB but just hangs. No errors no matter how many times I run. Definitely not getting by form_bronchitis but no indication why. 
The screenshots are deceiving because i'm sure I get by where it last updated screen!!!
This is occurring on my clinics production database but am unsure if a windows issue or maybe maria 10.1.19 i'm running local.

**Comment by sjpadgett** [comment]
@bradymiller 
$db_encoding = 'utf8mb4';
Just a bit ambitious don't you think? I don't see this working out too well or do you have a plan?

**Comment by bradymiller** [comment]
hi @sjpadgett , That should only be for new installs. Upgraders should retain their original sqlconf.php script and not be affected (in theory at least). And if go to a new sqlconf.php script (by copying old credentials into the new script), then the utf8mb4 should be changed to utf8 (or just removing the variable altogether).

**Comment by bradymiller** [comment]
btw, we used the same strategy a decade ago when going from latin1 to utf8, which is why that weird `$disable_utf8_flag = false;` variable is in sqlconf.php .

**Comment by sjpadgett** [comment]
duh! :) okay, thanks..

**Comment by bradymiller** [comment]
Is that what caused the sql upgrade error or are we still on the hunt for the issue ?

**Comment by sjpadgett** [comment]
Prob some of it but i'm getting ready to run it up on my linux box. we'll see...

**Comment by bradymiller** [comment]
I also just noted this issue on linux (docker dev environment when using demo data, which basically runs through the sql upgrade).

**Comment by bradymiller** [comment]
disregard my above post; my error was different and related to development in the branch i was working on. time to give myself a time out :)

**Comment by sjpadgett** [comment]
Yep having a dickens of a time with upgrade_sql just going away and sometimes will show an error mostly for unique indexes on the uuid stuff.

Have we set a min version for v6.0 upgrades?

**Comment by bradymiller** [comment]
Upgrade testing is working for me on the easy dev env docker when going from 5.0.0 (this is because the demo data option in devtools brings in a 5.0.0 database with demo data in it and then performs upgrade on it to most recent version, in this case 6.0.0-dev). This environment is using mariadb 10.5. The demo farm dev instances with demo data also do same thing and are using mariadb 10.3 (they are now resetting, so can't test them, but should be up anytime now for testing; when they are up the output for the upgrade should be found at https://eight.openemr.io/log/logSetup.txt note this is output for several demos so just need to look for all the upgrade sql in there).

**Comment by sjpadgett** [comment]
Appears the upgrade script is fine. My problem is the appearance that the upgrade hangs up due to the unusually long time it takes to create all the uuid infrastructure on a moderately sized production database.
On windows we are talking at least 15 minutes or more though, i've not timed it. I will time it on my linux box here soon but, in the end, we need to give an immediate feedback to the user as to where in the upgrade process we are. 
Because of output buffering etc currently the skipping or normal status echo's are of no help.

Maybe we grab the status from the server and display in a status bar at top of window...

![image](https://user-images.githubusercontent.com/1859985/91666627-57790b80-eacc-11ea-8fef-3604122d8c06.png)

**Comment by bradymiller** [comment]
Thinking that maybe an ajax element that checks something every 5 seconds or so (could send the output to a tmp log file placed in sites/default/documents/logs_and_misc/ also in addition to the screen (as already done which is basically impossible to control the buffering issues since can be out of individual apps scope)), which the ajax element will keep checking and showing output from). Also, as with we did with innodb conversion in past, need to have a good warning message that going to 6.0.0 can take a very long time. I'm happy to work on this.

**Comment by sjpadgett** [comment]
Thanks Brady, let me finish testing. Right now with mysql 5.7 for some reason it looks like the index exist check isn't working because script still tries to create the index. I can't put my finger on why my servers are having such a hard time with this upgrade.

The upgrade is not doing anything extraordinary unless there's a server(mysql) buffering problem or time outs i'm not seeing. 

I'm going to try a server stat check query to see if I can get an idea where issue is.

**Comment by sjpadgett** [comment]
@bradymiller 
There will be a lag in the server process status and the echoed query from upgrade script so, what do you think?
- maintain the entire server process history or
- just show last/current server event in history view.
here's scroll. still not styled...
![image](https://user-images.githubusercontent.com/1859985/91791385-027ae980-ebe1-11ea-8546-d58a733fd102.png)

**Comment by bradymiller** [comment]
hi @sjpadgett , Would show it all for now. Can get more fancy as we go (things that would be neat would be to expand/collapse the element with Show Details/Hide Details and show a very non-obtrusive ongoing indicator in the element until the script is completed). Since logging every single query wonder what impact of tens of thousands of queries would be (in terms of the write and also the ajax requests when the log file get really big); i am guessing we will see :)

**Comment by bradymiller** [comment]
oh and forgot to mention, the OpenEMR fans are very happy with this :)

![2DV](https://user-images.githubusercontent.com/278968/91795206-e3c12680-ebd1-11ea-98dc-1e42fea8b584.gif)

**Comment by sjpadgett** [comment]
lol, i've often wondered if the guy with one eye resents all the others with both eyes! And did he have to pay more to have his goggle made?

Nice thing about using the built in process list, it's maintained by the server and is a simple select query.
I have/had several options for checking status. Ajax, ES6 fetch w/promise, server side events or workers. Because this is a single process, polling with a promised fetch makes the best sense. Although, setting up workers may in the end, be a better option which I may still do depending how well current solution profiles.

This is the first time i've looked closely at this script and it really is a nicely done clever effort so, kudos to the author!

I'm going to clean up what I have and put up a PR so you can ponder(or marvel:)) over it and bring in for testing...

**Comment by sjpadgett** [comment]
@tywrenn 
Which is better to use: bg-light/dark text-light/dark or bg-primary/secondary etc?

**Comment by sjpadgett** [comment]
I can't seem to help myself:
![image](https://user-images.githubusercontent.com/1859985/91929579-ada4a500-ecac-11ea-80b9-bc62550c2c9c.png)
also fixed the < 5.0.0 warning to only display if needed.
![image](https://user-images.githubusercontent.com/1859985/91929902-71be0f80-ecad-11ea-8b15-1fb1d44bf4a9.png)

And translated most of upgrade.  I think i'll stop here and get back to some real work...

@tywrenn I found an issue with BS4 w-xxx and h-xxx classes where if the DOCBLOCK is html, percentages don't work. Only viewport ie 50vh etc. I've seen this in the past and could not figure out why I had to goto a viewport style. Now we know and I bet will drive folks crazy!
